RetryAnalyzer should work with a data provider

krmahadevan opened the pull request:
Closes #1946

Fixes #1946 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

juherr commented on src/main/java/org/testng/internal/Invoker.java:
```diff
@@ -1309,6 +1312,18 @@ private void handleInvocationResults(
+      String keyAsString = tm.getTestClass().getRealClass().getName() + "_" +
+          tm.getMethodName() + "_" + Arrays.toString(key);
+      retryAnalyzer = m_testMethodToRetryAnalyzer.computeIfAbsent(keyAsString,
+          o -> ClassHelper.newInstance(tm.getRetryAnalyzer().getClass()));
```
If I understand well the fix, we need a new instance by test method (or test result) instance?
Does it work if you map by TestResult instance (it will simplify the code)?

I wonder if `tm.getRetryAnalyzer()` is broken and should be deprecated. Then maybe the current logic could be moved somewhere else.

krmahadevan replied:
> Does it work if you map by TestResult instance (it will simplify the code)?

No it doesn't work.

TestNG already produces a new instance of `IRetryAnalyzer` per test method (`@Test` method). 

> we need a new instance by test method (or test result) instance?

What we need is a new instance of `IRetryAnalyzer` per parameter invocation for a given `@Test` method which is what this PR attempts at addressing.

The logic had to be housed in `Invoker` instead of `ITestNGMethod` because it is not aware of the parameters that are associated with it. That information is currently available only within a `ITestResult` object.

> I wonder `if tm.getRetryAnalyzer()` is broken and should be deprecated. 

I now added a new method that retrieves the `IRetryAnalyzer` from the `ITestNGMethod` when an `ITestResult` object is passed. I have amended the commit and pushed in the changes.

krmahadevan commented:
@juherr - Please take a look now. I have now relocated the logic back into `ITestNGMethod` implementation for computing the `IRetryAnalyzer` object that is to be used.

krmahadevan commented on src/main/java/org/testng/ITestNGMethod.java:
```diff
@@ -154,7 +154,17 @@
+  /**
+   * @deprecated - This method stands deprecated as of TestNG 7.0.0
+   */
+  @Deprecated
```
I didn't understand both these steps. Can you please elaborate ?

> With your solution, we miss an important part of interface : force method implementation

True that. I can get rid of the `default` implementation if that makes more sense, and thus force method implementation.

juherr replied:
The objective of such modification is to change api without breaking things: 
we want to use the new method, but we can't add it in interfaces (default method is the clue) and we can't remove the old method (because we use it in the default method).
So we just have to deprecate the old one in the first step.

Some versions later, we can assume the breaking change is known and we can remove the deprecated method and replace the default impl by the signature.

krmahadevan commented on src/main/java/org/testng/internal/TestNGMethod.java:
```diff
@@ -176,7 +176,7 @@ public BaseTestMethod clone() {
     clone.m_successPercentage = getSuccessPercentage();
     clone.setTimeOut(getTimeOut());
-    clone.setRetryAnalyzer(getRetryAnalyzer());
+    clone.setRetryAnalyzer(getRetryAnalyzer(null));
```
> Maybe we need a setRetryAnalyzerClass and getRetryAnalyzerClass?

That is going to add redundancy. The current design fetches the object from `org.testng.annotations.ITestAnnotation#getRetryAnalyzer` and then sets it within the `TestNGMethod` object via `org.testng.internal.TestNGMethod#init`. So if we end up adding a `setRetryAnalyzserClass()` and a corresponding `getRetryAnalzyserClass()` then are you proposing that we no longer get the instance via a call to `org.testng.annotations.ITestAnnotation#getRetryAnalyzer` but instead create the retry analyzer instance directly by using `ClassHelper.newInstance()` ? That would mean that we need to get rid of the method `org.testng.internal.BaseTestMethod#setRetryAnalyzer`. 

So in a nutshell 
1. Get rid of `org.testng.internal.BaseTestMethod#setRetryAnalyzer`
2. Introduce a `setRetryAnalyzerClass()` and `getRetryAnalyzerClass()` into `ITestNGMethod`

juherr replied:
Yes, it looks better. Can you try?

My last problem will be to understand why we need test result and determine if having information is test method will be better (agree, it means more changes).

krmahadevan commented:
@juherr - I have incorporated the changes based on our discussion. Please take a look.

krmahadevan commented:
@juherr - Request you to please take another look at this PR. The comments that you have added are on outdated changes. I am getting a bit confused with respect to what you are proposing.

juherr commented on src/main/java/org/testng/internal/BaseTestMethod.java:
```diff
@@ -743,4 +766,20 @@ public IParameterInfo getFactoryMethodParamsInfo() {
   }
+
+  private IRetryAnalyzer getRetryAnalyzerConsideringMethodParameteters(ITestResult tr) {
+    Object[] key = tr.getParameters();
```
My last problem is here because it is not really obvious why parameters are on the result and not on the method.

But if I remember well, TestMethod is something like the static description of the method (1 instance by method) and TestMethodInstance doesn't exist for the moment (1 instance by method run). So TestResult is the only way to catch parameters.

krmahadevan replied:
True. Parameter info is better off if its bundled with a test method. But then again we need to dig in deeper to figure out how to refactor that. We can do that in a separate PR.

krmahadevan commented on src/main/java/org/testng/internal/annotations/TestAnnotation.java:
```diff
@@ -22,6 +22,7 @@
   private boolean m_singleThreaded = false;
   private Class<?> m_dataProviderClass = null;
   private IRetryAnalyzer m_retryAnalyzer = null;
```
I will get rid of that attribute when the deprecated code path is removed. Its easy that way.

juherr replied:
But we don't need it anymore

krmahadevan replied:
@juherr - We need it to ensure backward compatibility. Third party clients who are extending `org.testng.internal.TestNGMethod` will still be traversing that codepath because our interface has new additions but those methods are `default` in nature.

There's nothing wrong in leaving it there and removing it off when the deprecated code paths are going to be removed no ?

krmahadevan commented on src/main/java/org/testng/internal/BaseTestMethod.java:
```diff
@@ -624,11 +629,29 @@ public IRetryAnalyzer getRetryAnalyzer() {
+  @Override
+  public void setRetryAnalyzerClass(Class<? extends IRetryAnalyzer> clazz) {
+    m_retryAnalyzerClass = clazz;
+    if (isRetryAnalyzerNotTestNGInjected(clazz)) {
```
fixed

krmahadevan commented on src/main/java/org/testng/ITestNGMethod.java:
```diff
@@ -154,10 +154,24 @@
+  @Deprecated
   void setRetryAnalyzer(IRetryAnalyzer retryAnalyzer);
 
+  IRetryAnalyzer getRetryAnalyzer(ITestResult result);
```
fixed.

krmahadevan commented:
@juherr - I have addressed all the review comments. Please take a look

juherr commented on src/main/java/org/testng/ITestNGMethod.java:
```diff
@@ -154,10 +154,30 @@
   void setRetryAnalyzer(IRetryAnalyzer retryAnalyzer);
 
+  default IRetryAnalyzer getRetryAnalyzer(ITestResult result) {
+    return null;
```
`return getRetryAnalyzer()` otherwise you'll have to check null value from here and then call previous getRetryAnalyzer in the business code (where we expect to call the new method only)

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/ITestNGMethod.java:
```diff
@@ -154,10 +154,30 @@
+  default void setRetryAnalyzerClass(Class<? extends IRetryAnalyzer> clazz) {}
+
+  default Class<? extends IRetryAnalyzer> getRetryAnalyzerClass() {
+    return null;
```
Here, you must return `getRetryAnalyzer().getClass()`

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/ITestNGMethod.java:
```diff
@@ -154,10 +154,30 @@
+    return null;
+  }
+
+  default void setRetryAnalyzerClass(Class<? extends IRetryAnalyzer> clazz) {}
```
Here you must call `setRetryAnalyzer(....newInstance(clazz))`

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/annotations/ITestAnnotation.java:
```diff
@@ -66,10 +66,19 @@
   void setRetryAnalyzer(Class<? extends IRetryAnalyzer> c);
 
+  default Class<? extends IRetryAnalyzer> getRetryAnalyzerClass() {
+    return null;
```
Here you must return `getRetryAnalyzer().getClass()`

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/internal/BaseTestMethod.java:
```diff
@@ -624,11 +629,27 @@ public IRetryAnalyzer getRetryAnalyzer() {
     return m_retryAnalyzer;
```
you should throw an exception if the method is called because it is not supposed to be the case

krmahadevan replied:
@juherr - That basically is as good as us removing the methods without deprecation. I believe the whole notion of deprecation is to just warn users that an API is going to be removed but still work. Throwing exceptions will cause the functionality to cease working and need code change by downstream users. If that's what we need then why use a deprecation strategy ? I am not sure I get this.

juherr commented on src/main/java/org/testng/internal/BaseTestMethod.java:
```diff
@@ -624,11 +629,27 @@ public IRetryAnalyzer getRetryAnalyzer() {
+  }
+
   @Override
   public void setRetryAnalyzer(IRetryAnalyzer retryAnalyzer) {
```
you should throw an exception if the method is called because it is not supposed to be the case

krmahadevan replied:
@juherr - That basically is as good as us removing the methods without deprecation. I believe the whole notion of deprecation is to just warn users that an API is going to be removed but still work. Throwing exceptions will cause the functionality to cease working and need code change by downstream users. If that's what we need then why use a deprecation strategy ? I am not sure I get this.

juherr commented on src/main/java/org/testng/internal/ClonedMethod.java:
```diff
@@ -147,6 +148,21 @@ public IRetryAnalyzer getRetryAnalyzer() {
     return m_method.getRetryAnalyzer();
```
you should throw an exception if the method is called because it is not supposed to be the case
Same for the setter

krmahadevan replied:
@juherr - That basically is as good as us removing the methods without deprecation. I believe the whole notion of deprecation is to just warn users that an API is going to be removed but still work. Throwing exceptions will cause the functionality to cease working and need code change by downstream users. If that's what we need then why use a deprecation strategy ? I am not sure I get this.

juherr commented on src/main/java/org/testng/internal/Invoker.java:
```diff
@@ -1281,8 +1281,13 @@ private void handleInvocationResults(
     boolean handled = holder.handled;
+    IRetryAnalyzer retryAnalyzer = testMethod.getRetryAnalyzer(testResult);
+
+    //For backward compatibility.
```
What I said previously: if the default method is well implemented, you don't have to do that ;)

krmahadevan replied:
True that. Fixed.

juherr commented on src/main/java/org/testng/internal/WrappedTestNGMethod.java:
```diff
@@ -274,6 +275,21 @@ public void setRetryAnalyzer(IRetryAnalyzer retryAnalyzer) {
     testNGMethod.setRetryAnalyzer(retryAnalyzer);
```
you should throw an exception if the method is called because it is not supposed to be the case
same for getter

krmahadevan replied:
@juherr - That basically is as good as us removing the methods without deprecation. I believe the whole notion of deprecation is to just warn users that an API is going to be removed but still work. Throwing exceptions will cause the functionality to cease working and need code change by downstream users. If that's what we need then why use a deprecation strategy ? I am not sure I get this.

krmahadevan commented:
@juherr - I have fixed some of the comments. For the comments which needed an exception to be thrown, I have added my comments for them (I feel it will break backward compatibility if a deprecated method basically starts throwing exceptions). I have also rebased off of master to resolve merge conflicts. Please take a look.

kdevendraraju commented:
@krmahadevan 

Can you please help me to understand, below two queries on EmailableReporter2.java

1. Why are we not writing scenario details for retried tests
2. Why are we just writing the timing information with the first scenario per method. If we consider retry case, we have different timing information for different test.

Thank You.
